### PR TITLE
Create laws and tests: Abbe's invariant, formula of spherical refractive surface

### PR DESCRIPTION
**Pull request by Voldoned**

### CODE REVIEW

File: symplyphysics/laws/optics/abbe_invariant.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+from sympy import (Eq, solve)
+from symplyphysics import (units, Quantity, Symbol, print_expression,
+    validate_input, validate_output, dimensionless)
+
+# Description
+## The Abbe invariant is an optical characteristic of the refractive curve of the surface
+
+# Law: Q = n * ((1 / d) - (1 / R))
+# Where:
+## n - surface refraction index
```
Comment (blackyblack) [inline symplyphysics/laws/optics/abbe_invariant.py:10]: It's not surface refraction index. It's medium refraction index before or after the surface.

File: symplyphysics/laws/optics/abbe_invariant.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+## d - distance to surface
+## R - surface curvature radius
```
Comment (blackyblack) [inline symplyphysics/laws/optics/abbe_invariant.py:12]: Does the sign change for convex and concave surface?

File: symplyphysics/laws/conservation/abbe_invariant_of_two_optical_environments_is_constant.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+from sympy import Eq, solve
+from symplyphysics import (Symbol, units, print_expression, Quantity,
+    validate_input, validate_output)
+
+# Description
+## The Abbe's invariant connects the front and back segments S and S',
+## allowing one of them to be determined if the second one is known
+
+# Law: Q = Q'
```
Comment (blackyblack) [inline symplyphysics/laws/conservation/abbe_invariant_of_two_optical_environments_is_constant.py:9]: Maybe it's better to combine both Abbe invariant definition and it's conservation to a single law? This way it is easier to understand, what's going on and no need to introduce Abbe invariant that does not have physical meaning.

File: symplyphysics/laws/optics/abbe_invariant.py (hunk 1 above)
Comment (Voldoned) [reply symplyphysics/laws/optics/abbe_invariant.py:10]: You're right. I wanted to say: "the refractive index of the optical environment"

File: symplyphysics/laws/optics/abbe_invariant.py (hunk 2 above)
Comment (Voldoned) [reply symplyphysics/laws/optics/abbe_invariant.py:12]: It was meant that the surface into which the ray enters is convex. I will indicate this at the next fix.

File: symplyphysics/laws/conservation/abbe_invariant_of_two_optical_environments_is_constant.py (hunk 3 above)
Comment (Voldoned) [reply symplyphysics/laws/conservation/abbe_invariant_of_two_optical_environments_is_constant.py:9]: Initially, I also thought of drafting one law, but then I came across such a task:

"A thin plano-convex glass lens with a refractive index of n2 = 1.66, the convex side with a radius of curvature R = 12 cm is immersed in water. The refractive index of water is n1=1.33. At what distance from the lens and where will the image of the Sun be located at the zenith?"

image

And I thought that if the beam passes through three media (for example, air, biconvex lens and water), then we will need an Abbe invariant for each medium, although I did not find such difficult tasks.

Comment (Voldoned) [reply symplyphysics/laws/conservation/abbe_invariant_of_two_optical_environments_is_constant.py:9]: I found an additional explanation:

"The Abbe formula is valid only when passing through one refractive surface, with a change in the position of the conjugate points, the value of the Abbe invariant will change, therefore it is not a complete invariant."

Therefore, yes, it is necessary to redo it into one law.

File: symplyphysics/laws/conservation/abbe_invariant_of_two_optical_environments_is_constant.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+from sympy import Eq, solve
+from symplyphysics import (Symbol, units, print_expression, Quantity,
+    validate_input, validate_output)
+
+# Description
+## The Abbe's invariant connects the front and back segments S and S',
+## allowing one of them to be determined if the second one is known
```
[the remark this answers is not in the document]
Comment (Voldoned) [reply symplyphysics/laws/conservation/abbe_invariant_of_two_optical_environments_is_constant.py:7]: It's about the optical axis. The point S is located on the front of the optical axis, i.e. on the part that is outside the spherical lens (outside). The point S' is located on a part of the optical axis inside the lens

File: symplyphysics/laws/conservation/abbe_invariant_of_two_optical_environments_is_constant.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                           convert_to)
+
+# Description
+## The Abbe's invariant connects the front and back segments S and S',
+## allowing one of them to be determined if the second one is known
+
+# Law: n * ((1 / a) - (1 / R)) = n' * ((1 / b) - (1 / R))
+# Where:
+## n - refraction index in optical environment
+## n' - refraction index in lens
```
Comment (blackyblack) [inline symplyphysics/laws/conservation/abbe_invariant_of_two_optical_environments_is_constant.py:13]: Somehow you limit it to lens, but this law is about any surface with constant curvature radius, between 2 optical environments.

File: symplyphysics/laws/optics/the_formula_of_the_spherical_refractive_surface.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
+from sympy import (Eq, solve, S)
```
Comment (blackyblack) [inline symplyphysics/laws/optics/the_formula_of_the_spherical_refractive_surface.py:1]: Remove `the formula` from file name. Use format `what_we_get_from_what_we_put`

File: symplyphysics/laws/optics/the_formula_of_the_spherical_refractive_surface.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+law = Eq(
+    (refraction_index_environment / (-distance_to_object)) + (refraction_index_lens / distance_to_image),
+    (refraction_index_lens - refraction_index_environment) / curvature_radius_lens
+)
+
+# From Abbe's invariants:
+
+abbe_invariant_environment = refraction_index_environment * ((1 / distance_to_object) - (1 / curvature_radius_lens))
+abbe_invariant_lens = refraction_index_lens * ((1 / distance_to_image) - (1 / curvature_radius_lens))
```
Comment (blackyblack) [inline symplyphysics/laws/optics/the_formula_of_the_spherical_refractive_surface.py:39]: Use law for it, don't copypaste.

File: symplyphysics/laws/optics/the_formula_of_the_spherical_refractive_surface.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+# Where:
+## n - medium refraction index
+## lo - distance from object to surface
+## li - distance from image to surface
+## R - surface curvature radius
+
+# NOTES:
+## - minus before lo if the origin of the coordinate system is placed at the intersection of the optical axis
+##   and the surface of the lens with the radius lying on this optical axis;
+## - proofs: https://studme.org/341451/matematika_himiya_fizik/prelomlenie_otrazhenie_sveta_sfericheskoy_poverhnosti.
```
Comment (blackyblack) [inline symplyphysics/laws/optics/the_formula_of_the_spherical_refractive_surface.py:22]: Somehow it looks like this formula explains the image position of the lens, but it is not totally true. In normal lens there are 2 refractive surfaces, so Abbe's invariant should be applied twice. In this proof the focal point is somewhere inside the lens, so only one refractive surface is being calculated. Please mention this fact somewhere in the description.

File: test/conservation/abbe_invariant_of_two_optical_environments_is_constant_test.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    SI, prefixes,
+)
+from symplyphysics.laws.conservation import abbe_invariant_of_two_optical_environments_is_constant as abbe_conservation_law
+
+# A thin, flat-convex lens is slightly immersed in water
+# with its horizontal flat side (the convex surface of the lens is in the air).
+# A narrow vertical beam of light falls on the lens from above,
+# the axis of which passes exactly through the top of the convex surface.
+# This beam is focused in water at a depth of h = 27 cm. The optical power of the lens
+# in the air is D = 5 dptr. Find the refractive index of water.
```
Comment (blackyblack) [inline test/conservation/abbe_invariant_of_two_optical_environments_is_constant_test.py:17]: It is like a full example. Do we have something easier for a test?

File: test/conservation/abbe_invariant_of_two_optical_environments_is_constant_test.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    a = oo
+    b = Quantity(27 * prefixes.centi * units.meters)
+
+    Args = namedtuple("Args", ["a", "b", "n", "r"])
+    return Args(a=a, b=b, n=n, r=r)
+
+
+def test_basic_conservation(test_args):
+    result = abbe_conservation_law.calculate_refraction_index_lens(test_args.a, test_args.b, test_args.n, test_args.r)
+    assert SI.get_dimension_system().equivalent_dims(result.dimension, 1 / units.length)
```
Comment (blackyblack) [inline test/conservation/abbe_invariant_of_two_optical_environments_is_constant_test.py:38]: I thought refraction index is dimensionless.

File: test/conservation/abbe_invariant_of_two_optical_environments_is_constant_test.py [hunk 11]
```diff
@@ -26,30 +26,28 @@
...
+        abbe_conservation_law.calculate_refraction_index_lens(db, test_args.b, test_args.r, test_args.n)
     with raises(TypeError):
-        abbe_conservation_law.calculate_refraction_index_lens(100, test_args.b, test_args.n, test_args.r)
+        abbe_conservation_law.calculate_refraction_index_lens(100, test_args.b, test_args.r, test_args.n)
 
     with raises(errors.UnitsError):
-        abbe_conservation_law.calculate_refraction_index_lens(test_args.a, db, test_args.n, test_args.r)
+        abbe_conservation_law.calculate_refraction_index_lens(test_args.a, db, test_args.r, test_args.n)
     with raises(TypeError):
         abbe_conservation_law.calculate_refraction_index_lens(test_args.a, 100, test_args.n, test_args.r)
```
Comment (blackyblack) [inline test/conservation/abbe_invariant_of_two_optical_environments_is_constant_test.py:52]: Do you want to fix this as well?

**Review by blackyblack — APPROVED** [review]

File: symplyphysics/laws/optics/the_formula_of_the_spherical_refractive_surface.py (hunk 6 above)
Comment (blackyblack) [reply symplyphysics/laws/optics/the_formula_of_the_spherical_refractive_surface.py:1]: Not fixed.

**Review by blackyblack — APPROVED** [review]

**Pull request merged**